### PR TITLE
Adding restart for non converging Krylov algorithm

### PR DESCRIPTION
**Pull request by kbidzhiev**
Summary

This PR updates the Lanczos implementation to use a residual-minimizing restart strategy for DMRG only. After each iteration the algorithm evaluates the residuals of the candidate vectors, selects the vector with the smallest residual, and restarts the search from that vector.

What changed

- Added residual-based selection across the current set of candidate vectors.
- Modified the restart logic so the next Lanczos run begins from the best (lowest-residual) vector instead of continuing from the last generated vector.
- Updated the iteration flow to support “pick-best-and-restart” without altering the external API/inputs.

exhaustive guide to numerical errors etc in Lanczos algo 
Lectures on Lanczos with detail study of residual
https://www.stat.uchicago.edu/~lekheng/courses/302/demmel/demmch7.pdf

### CODE REVIEW

File: emu_base/math/krylov_energy_min.py [hunk 1]
```diff
@@ -4,104 +4,187 @@
...
-                c * vec for c, vec in zip(ground_eigenvector, lanczos_vectors)
-            )
-            final_state = final_state / final_state.norm()
-            happy_breakdown = True
-            converged = True
-            break
 
-        # Reconstruct final state in original Hilbert space
-        lanczos_vectors.append(w / beta)
-        final_state = sum(c * vec for c, vec in zip(ground_eigenvector, lanczos_vectors))
```
Comment (kbidzhiev) [inline emu_base/math/krylov_energy_min.py:89]: size of `ground_eigenvector` is old, but `lanczos_vectors` are updated.
now they have diff size

File: emu_base/math/krylov_energy_min.py [hunk 2]
```diff
@@ -1,121 +1,227 @@
...
+    v = torch.tensordot(c, Q, dims=([0], [0]))
 
-def krylov_energy_minimization_impl(
+    norm = v.norm()
+    if norm.item() <= NUMERICAL_TOLERANCE:
+        raise ValueError("Ritz vector has zero norm")
+    return cast(torch.Tensor, v / norm)  # mypy requires explicit type
+
+
+def build_next_lanczos_vector(
```
Comment (a-quelle-pasqal) [inline emu_base/math/krylov_energy_min.py]: The name of this function is not descriptive of what it actually does, which is building the next lanczos vector AND filling the T matrix

File: emu_base/math/krylov_energy_min.py [hunk 3]
```diff
@@ -1,121 +1,227 @@
...
+def _ritz_vector(coefficients: torch.Tensor, basis: list[torch.Tensor]) -> torch.Tensor:
+    """
+    Return the normalized Ritz vector from Lanczos basis vectors.
+    """
+    if len(coefficients) != len(basis):
+        raise ValueError(
+            f"Expected len(coefficients) == len(basis), got {len(coefficients)} != {len(basis)}"
+        )
 
+    Q = torch.stack(basis)
```
Comment (a-quelle-pasqal) [inline emu_base/math/krylov_energy_min.py]: I dislike this, it duplicates all the basis vectors in memory, which can mean a lot of memory is wasted, since storing the Lanczos vectors is where most of the memory in the algorithm is used. The original implementation with `sum` was better.

File: emu_base/math/krylov_energy_min.py (hunk 1 above)
Comment (a-quelle-pasqal) [reply emu_base/math/krylov_energy_min.py:89]: zip only goes as far as the shortes of the two lists. I think they were always of different size, and that is ok?

**Review by a-quelle-pasqal — CHANGES_REQUESTED** [review]
Please also add the literature, I'm especially interested in the justification for
`resid = (beta * y[-1]).abs()  # == norm(op(ritz_vec) - ritz_value * ritz_vec)`

File: test/emu_base/math/test_krylov_energy_min.py [hunk 4]
```diff
@@ -201,12 +205,120 @@ def test_live_data(lb, rb, combined_hamiltonian_factors, v, krylov_dim):
...
+    expected_e0 = torch.tensor(-1.0)
+    expected_v0 = torch.tensor([1.0, -1.0]) / torch.sqrt(torch.tensor(2))
+    sgn = expected_v0 @ v0
+    expected_v0 *= sgn
+    assert torch.allclose(e0, expected_e0)
+    assert torch.allclose(v0, expected_v0)
+
+
+def test_lowest_eigen_pair_rejects_non_hermitian():
+    h = torch.tensor([[1.0, 2.0], [0.0, 1.0]])
```
Comment (a-quelle-pasqal) [inline test/emu_base/math/test_krylov_energy_min.py]: I'm not really sure we need the asserts in `lowest_eigen_pair` an din `ritz_vector`. We have full control over what goes into those functions, and it shouldn't ever trigger those checks, while they do take time to compute.

File: emu_base/math/krylov_energy_min.py (hunk 3 above)
Comment (kbidzhiev) [reply emu_base/math/krylov_energy_min.py]: 🚀

File: emu_base/math/krylov_energy_min.py (hunk 1 above)
Comment (kbidzhiev) [reply emu_base/math/krylov_energy_min.py:89]: They must be the same size, since this is a dot product.
My comment on the old code is meant to highlight our previous misunderstanding of how the next vector is built and then it updates lanczos basis

File: test/emu_base/math/test_krylov_energy_min.py (hunk 4 above)
Comment (kbidzhiev) [reply test/emu_base/math/test_krylov_energy_min.py]: refactored and removed tests

We previously ran into an issue where T_matrix was silently not Hermitian. I now build tridiag explicitly and kept the comment as a reminder for potential future use with LOBPCG or other methods.

File: emu_base/math/krylov_energy_min.py [hunk 5]
```diff
@@ -1,121 +1,227 @@
...
+    return torch.vdot(a.reshape(-1), b.reshape(-1))
+
+
+def _lowest_eigen_pair(h: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
     """
-    Return the lowest eigenpair of the hermitian matrix T_trunc.
+    Return the lowest eigenpair of the hermitian matrix h.
     """
-    eig_energy, eig_state = torch.linalg.eigh(T_trunc)
+    err = (h - h.mH).norm()
```
Comment (elmerhej) [inline emu_base/math/krylov_energy_min.py]: why don't you `torch.allclose(h, h.T)` instead? since the truncated matrix is built symmetrically anyway

File: emu_base/math/krylov_energy_min.py [hunk 6]
```diff
@@ -1,121 +1,227 @@
...
+    ground_state: torch.Tensor
+    ground_energy: float
+    residual_norm: float
+    converged: bool
+    happy_breakdown: bool
+    iteration_count: int
+    restart_count: int
+
+
+def _dotc(a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
```
Comment (elmerhej) [inline emu_base/math/krylov_energy_min.py]: I think this function is redundant

File: emu_base/math/krylov_energy_min.py (hunk 2 above)
Comment (kbidzhiev) [reply emu_base/math/krylov_energy_min.py]: fair. done

File: emu_base/math/krylov_energy_min.py (hunk 5 above)
Comment (kbidzhiev) [reply emu_base/math/krylov_energy_min.py]: refactored. the `h` matrix is build "on fly"

**Comment by kbidzhiev** [comment]
> Please also add the literature, I'm especially interested in the justification for `resid = (beta * y[-1]).abs() # == norm(op(ritz_vec) - ritz_value * ritz_vec)`

see Saad (1), Prop. 6.8 (Ch. 6, p. 131).

(1) Numerical Methods for Large Eigenvalue Problems, Y. Saad.
https://epubs.siam.org/doi/book/10.1137/1.9781611970739

File: emu_base/math/krylov_energy_min.py (hunk 6 above)
Comment (kbidzhiev) [reply emu_base/math/krylov_energy_min.py]: it's just a wrapper, I tried to avoid explicit reshape and conjugation for the `lhs` since `torch.vdot` does it automatically

Comment (kbidzhiev) [reply emu_base/math/krylov_energy_min.py]: removed

**Review by elmerhej — DISMISSED** [review]

File: test/emu_base/math/test_krylov_energy_min.py
Comment (a-quelle-pasqal) [inline test/emu_base/math/test_krylov_energy_min.py]: could you check that these tests now converge for a lower max krylov dim? Say 50?

File: test/emu_base/math/test_krylov_energy_min.py (hunk 4 above)
Comment (a-quelle-pasqal) [reply test/emu_base/math/test_krylov_energy_min.py]: The comment is good, but the asserts are still in the code. Are you arguing for also keeping them?

File: emu_base/math/krylov_energy_min.py [hunk 7]
```diff
@@ -1,121 +1,242 @@
...
+    Restart strategy (explicit):
+      - Run a Lanczos cycle up to `max_krylov_dim`.
+      - If not converged, restart with the best Ritz vector (lowest residual) from that cycle.
+
+    Convergence:
+      - residual norm ||Hψ - Eψ|| < residual_tolerance
+      - happy breakdown: beta < norm_tolerance
+    """
+
+    result = KrylovEnergyResult(
```
Comment (a-quelle-pasqal) [inline emu_base/math/krylov_energy_min.py]: this is not used? There is result returned in line `212`

File: emu_base/math/krylov_energy_min.py [hunk 8]
```diff
@@ -1,121 +1,242 @@
...
+        total_iters += result.iteration_count
+        result = replace(result, restart_count=r, iteration_count=total_iters)
+
+        if result.happy_breakdown or result.converged:
+            break
+
+    return result
+
+
 def krylov_energy_minimization(
```
Comment (a-quelle-pasqal) [inline emu_base/math/krylov_energy_min.py]: I had a discussion with Elie, and it would make the file easier to read if you reverse the order in which the functions are defined in the file. The entry point here will be `krylov_energy_minimization` so if you put this at the top, it will be where the user starts reading. Then they will see it calls `krylov_energy_minimization_impl` so this can be the function below in the file etc.

File: test/emu_base/math/test_krylov_energy_min.py (hunk 4 above)
Comment (kbidzhiev) [reply test/emu_base/math/test_krylov_energy_min.py]: removed

File: test/emu_base/math/test_krylov_energy_min.py
Comment (kbidzhiev) [reply test/emu_base/math/test_krylov_energy_min.py]: done

File: emu_base/math/krylov_energy_min.py (hunk 7 above)
Comment (kbidzhiev) [reply emu_base/math/krylov_energy_min.py]: I create a variable outside of the scope, so the loop look this way 

```
res = 0
for _ in range(..):
    new_res = update_res(res)
    res = new_res
```

File: emu_base/math/krylov_energy_min.py (hunk 8 above)
Comment (kbidzhiev) [reply emu_base/math/krylov_energy_min.py]: done

**Comment by kbidzhiev** [comment]
> Please also add the literature, I'm especially interested in the justification for `resid = (beta * y[-1]).abs() # == norm(op(ritz_vec) - ritz_value * ritz_vec)`

done in the comments in the code and in the head of the file

File: emu_base/math/krylov_energy_min.py (hunk 8 above)
Comment (elmerhej) [reply emu_base/math/krylov_energy_min.py]: Thanks for making the change :+1:

**Comment by elmerhej** [comment]
LGTM

**Review by a-quelle-pasqal — APPROVED** [review]
LGTM
